R/db-odbc-access.R

```diff
@@ -0,0 +1,240 @@
+                   substr        = function(x, start, stop){
+                     right  <- stop - start + 1
+                     left   <- stop
+                     build_sql(
+                       "RIGHT(LEFT(", x, ", ", left, "), ", right, ")")
+                   },
+                   trimws        = sql_prefix("TRIM"),
+                   # No support for CONCAT in Access
+                   paste         = sql_not_supported("paste()"),
+                   paste0        = sql_not_supported("paste0()"),
+
+                   # Logic
+                   # ISNULL() returns -1 for True and 0 for False
+                   # Override this by returning 1 for True and 0 for False
+                   is.null       = function(x){
+                     build_sql("IIF(ISNULL(", x ,"), 1, 0)")
+                   },
+                   is.na         = function(x){
+                     build_sql("IIF(ISNULL(", x ,"), 1, 0)")
+                   },
+                   # IIF() is like ifelse()
+                   ifelse        = function(test, yes, no){
+                     build_sql("IIF(", test, ", ", yes, ", ", no, ")")
+                   },
+
+                   # Coalesce doesn't exist in Access.
+                   # NZ() only works while in Access, not with the Access driver
+                   # IIF(ISNULL()) is the best way to construct this
+                   coalesce      = function(x, y) {
+                     build_sql("IIF(ISNULL(", x, "),", y, ",", x, ")")
+                   },
+
+                   # pmin/pmax for 2 columns
+                   pmin          = function(x, y) {
+                     build_sql("IIF(", x, " <= ", y, ",", x, ",", y, ")")
+                   },
+
+                   pmax          = function(x, y) {
+                     build_sql("IIF(", x, " <= ", y, ",", y, ",", x, ")")
+                   },
+
+                   # Dates
+                   Sys.Date      = sql_prefix("DATE")
+
+    ),
+
+    sql_translator(.parent = base_odbc_agg,
+
+                   mean          = sql_prefix("AVG"),
+                   sd            = sql_prefix("STDEV"),
+                   var           = sql_prefix("VAR"),
+                   max           = sql_prefix("MAX"),
+                   min           = sql_prefix("MIN"),
+                   # Access does not have functions for cor and cov
+                   cor           = sql_not_supported("cor()"),
+                   cov           = sql_not_supported("cov()"),
+
+                   # Count
+                   # Count(Distinct *) does not work in Access
+                   # This would work but we don't know the table name when translating:
+                   # SELECT Count(*) FROM (SELECT DISTINCT * FROM table_name) AS T
+                   n_distinct    = sql_not_supported("n_distinct")
+    ),
+
+    # Window functions not supported in Access
+    sql_translator(.parent = base_odbc_win,
```

Comment:
I think you can just use `base_no_win` here

Reply:
I have changed it to `base_no_win` and removed everything else, but am now getting:

```r
Warning message:
Translator is missing window functions:
cor, count, cov 
```

They don't exist in `base_no_win`. Should they?

Reply:
Also it seems this isn't entirely true. `Avg`, `Count`, `Min`, `Max`, and `Sum` are [supported aggregate functions](https://www.techonthenet.com/access/functions/numeric/avg.php) but use `GROUP BY` in the sql generated:

```r
dbGetQuery(con, "SELECT cyl, avg(`carb`) AS avg_carb FROM mtcars GROUP BY cyl")
```

This works fine. Note that in the `SELECT` you can only include columns that are either in the `GROUP BY` or wrapped in one of the 5 aggregate functions. So this with `mpg` would not work:

```r
dbGetQuery(con, "SELECT cyl, mpg, avg(`carb`) AS avg_carb FROM mtcars GROUP BY cyl")
```

I can see that `mtcars %>% group_by(cyl) %>% mutate(mean(carb))` would produce SQL with `OVER (PARTITION ...)` like this:

```r
SELECT `mpg`, `cyl`, `disp`, `hp`, `drat`, `wt`, `qsec`, `vs`, `am`, `gear`, `carb`, avg(`carb`) OVER (PARTITION BY `cyl`) AS `mean(carb)`
FROM `mtcars`
```

Is there an easy way to change that to use the Group By syntax?

Reply:
I don't think you need to supply `count()`. 

Can you please add `cov()` and `cor()` to `base_no_win`.

Reply:
After adding `cov()` and `cor()` the `count()` warning went away as well 👍